drainer/translator/translator.go

```diff
@@ -44,28 +44,50 @@ func getParser() (p *parser.Parser) {
 	return
 }
 
-func insertRowToDatums(table *model.TableInfo, row []byte) (pk types.Datum, datums map[int64]types.Datum, err error) {
+func insertRowToDatums(table *model.TableInfo, row []byte) (datums map[int64]types.Datum, err error) {
 	colsTypeMap := util.ToColumnTypeMap(table.Columns)
 
-	// decode the pk value
-	var remain []byte
-	remain, pk, err = codec.DecodeOne(row)
-	if err != nil {
-		return types.Datum{}, nil, errors.Trace(err)
+	var (
+		commonPKInfo *model.IndexInfo
+		pkLen        = 1
+	)
+	if table.IsCommonHandle {
+		for _, idx := range table.Indices {
+			if idx.Primary {
+				commonPKInfo = idx
+				break
+			}
+		}
+		if commonPKInfo != nil {
+			pkLen = len(commonPKInfo.Columns)
+		}
```

Comment:
What if `commonPKInfo` is `nil`? Although it seems impossible but we can still put a check here?

Reply:
yes~it's impossible that a clustered index table without pk,  add a check and return an error at https://github.com/pingcap/tidb-binlog/pull/1040/files#diff-d362ef1721274a90486b677b50c57cfec86fac0f7fcf567587b452fe7611898dR61 to avoid someone do it in future - -